### ISSUE TITLE
Cinnamon teamviewer issues

### ISSUE THREAD

**Issue by bawasaab** [linuxmint/Cinnamon#5508]
As i share my screen to my friend using team viewer, i was unable to access my menu, desktop as well as the minimize, maximize and close buttons of team viewer. It happens each and every time. So I were unable to share any file to him.

**Comment by feren** [comment]
Could you use File Transfer?
If so, my issue up above will be different to this one
Otherwise, if you're talking about the screen glitches, then I'll proceed with adding bug information...

**Comment by bawasaab** [comment]
yes please am waiting for your bug information
 

    On Monday, September 19, 2016 11:31 PM, feren <notifications@github.com> wrote:
 

 Could you use File Transfer?
If so, my issue up above will be different to this one
Otherwise, if you're talking about the screen glitches, then I'll proceed with adding bug information...—

**Comment by feren** [comment]
Here's some info:
- Does not show glitches on real machine
- Have to move the mouse enough to temporarily get rid of the glitches
- Not rare
- Did Not Happen Before 3.0

Steps Of Reproduction:
1. Get a victim machine with Cinnamon 3.0.X to install TeamViewer onto
2. Get a host machine to install TeamViewer onto as well (we'll need it for connecting to the 'victim' machine)
3. With the Host Machine, once you've got the Trusted Devices thing sorted out, Connect to the Victim machine from the host
4. Try to use Cinnamon and see how much you see the bug

Affects:
- Cinnamon 3.0.X
- Linux Mint 18 and pretty much every Ubuntu distro you can name that has Cinnamon 3.0 on it one way or another
- TeamViewer

**Comment by feren** [comment]
@JosephMcc UPDATE: It's not only TeamViewer affected
https://www.youtube.com/watch?v=S6RXuDhYeiA&feature=autoshare

**Comment by feren** [comment]
Update From Clem:
> That’s an issue with drivers and xorg

**Comment by feren** [comment]
@JosephMcc Hmm, it looks like Budgie suffers from a similar bug:
https://youtu.be/FeD4oERBrQI?t=10m9s

(Look closely at the panel and you'll see a similar bug)

**Comment by feren** [comment]
@JosephMcc I've noticed that Recording the Screen of a Cinnamon session also has the same bug on the outputted video as TeamViewer does on the Remote Controller's Screen, shall I open a new issue, or shall we also use this issue to address that bug too?

**Comment by leigh123linux** [comment]
Has anyone filed an issue against teamviewer?

**Comment by feren** [comment]
@leigh123linux Not yet, though I know that the exact same bug is part of Screen Recording Output, so I'll try recording Cinnamon in Software Rendering...

**Comment by feren** [comment]
Hang on, watching this video of KDE Neon, looks like this could actually be a bug with X.org itself...
Desktops Affected:
Cinnamon (Linux Mint and co.)
Budgie (Bubuntu aka Ubuntu Budgie)
KDE (KDE Neon)

Video: https://www.youtube.com/watch?v=ajj8iA4pRFw

**Comment by feren** [comment]
@leigh123linux Ok, looks like it's only happening on Non-Software Rendering Mode...
With Software Rendering Mode: https://www.dropbox.com/s/artr809hwuzr0jd/Cinnamon%20Software%20Rendering%20Output%20Test.mp4?dl=0
Without Software Rendering Mode (notice the screen glitches every now and then? They get worse other times): https://www.dropbox.com/s/wo0s0iy9tbf8ov7/Cinnamon%20Software%20Rendering%20Output%20Test%20Without%20Software.mp4?dl=0

**Comment by mainmachine** [comment]
OK, so it looks like the issue is down to Teamviewer and "Hardware Acceleration" - are these the two common factors?

I wouldn't jump to the conclusion that it's a bug with xorg, unless this happens with applications other than Teamviewer...

**Comment by mainmachine** [comment]
@feren - what GPU are you using here, and what driver in the case of Nvidia or AMD (Mesa or proprietary)?

Have you experienced this (or even tested it) on a variety of hardware?

**Comment by feren** [comment]
@mainmachine It happens on my Toshiba and HP, it's an Intel one and it is non-proprietary, it also affects Screen Recording Output as well

**Comment by feren** [comment]
@mainmachine
![screenshot from 2016-11-29 18-24-39](https://cloud.githubusercontent.com/assets/11057934/20723108/1d4c452e-b661-11e6-81e1-ae9736fcaac4.png)

**Comment by mainmachine** [comment]
OK, so on only Intel HW then? Or are you using something like Prime to use that Radeon GPU...?

I'm trying to isolate whether this is unique to one make of GPU hardware, and by proxy, one driver.

**Comment by feren** [comment]
@mainmachine Wow, I didn't even notice that AMD Graphics Card, hang on, let me run that GFX command in Terminal...

**Comment by feren** [comment]
@mainmachine
```
glxinfo | grep OpenGL
OpenGL vendor string: Intel Open Source Technology Center
OpenGL renderer string: Mesa DRI Intel(R) Sandybridge Mobile 
OpenGL core profile version string: 3.3 (Core Profile) Mesa 11.2.0
OpenGL core profile shading language version string: 3.30
OpenGL core profile context flags: (none)
OpenGL core profile profile mask: core profile
OpenGL core profile extensions:
OpenGL version string: 3.0 Mesa 11.2.0
OpenGL shading language version string: 1.30
OpenGL context flags: (none)
OpenGL extensions:
OpenGL ES profile version string: OpenGL ES 3.0 Mesa 11.2.0
OpenGL ES profile shading language version string: OpenGL ES GLSL ES 3.00
OpenGL ES profile extensions:
```

**Comment by mainmachine** [comment]
Here's another data point to look at:

I just tested two scenarios, using my laptop (Intel/Nvidia - Optimus) and an AMD APU (CPU/GPU) workstation.

I tested in each direction, i.e. each one as client and server. 

With the AMD workstation as server, I had no problems other than the expected lag using a utility like this - none of the glitchy, texture "disappearing" that has been reported here.

In the other direction, I was able to easily reproduce the problem. 

Let's be clear about the Intel/Nvidia setup on my laptop - Optimus has the Intel GPU connected to the display directly, so even if you're using the Nvidia GPU, it's passing it's output to the Intel Chip, so you are never outputting from the Nvidia GPU directly to the display (unless you use the HDMI port, but I don't have that configured in a way I can easily test for this)... 

So basically I am seeing this on Intel GPU, but not on AMD, and I would wager it wouldn't happen on an Nvidia GPU.

I use the Mesa/Radeon drivers on the AMD system, Mesa (only option) on Intel, and proprietary on Nvidia.

So from what I see here, Intel hardware (and thus driver stack) is the only place I'm seeing this bug.

So I'd guess there's something in the Mesa/Intel stack that isn't playing nice with whatever Cinnamon uses for that screen recording function - that points more toward a GTK/ Xorg issue with the way Mesa handles framebuffer access, or something to that effect (my core competencies don't include a high level of detail on the Mesa/driver stack)...

**Comment by mainmachine** [comment]
@feren - as an aside, I would seriously look at either ensuring the GPU is disabled (for power saving) or actually enabling the AMD GPU on that laptop - you'll either extend battery life or get better graphical performance, whichever you fancy.

I use Bumblebee on my Acer (Intel/Nvidia) and this allows me to do both with minimal PITA.

**Comment by mainmachine** [comment]
Oh, and ***bug confirmed***...

Steps to reproduce:
1. set up teamviewer on two machines, target machine should be running Intel GPU and not running in software rendering mode
2. remote access target Intel machine from the other 'un
3. behold the glitches

**Comment by feren** [comment]
You may want to tag @JosephMcc into that...

**Comment by mainmachine** [comment]
I think you just did. :)

**Comment by leigh123linux** [comment]
I doubt there is much we can do to fix the issue, all 3D De's suffer from it.

**Comment by leigh123linux** [comment]
@mainmachine  All you have proved is VNC  can't run 3D desktops, this isn't a bug in cinnamon.

**Comment by mainmachine** [comment]
@leigh123linux - not exactly, but I agree the problem is not unique to Cinnamon, and this bug should be closed. I didn't mean by my "confirmed" comment to attribute a cause, just that the bug is real and reproducible.

@feren - as @leigh123linux suggests, this needs to be explored more to identify where the source of the problem is, and the evidence is pointing to something at lower level than Cinnamon. I suspect either GTK level or even Mesa or Xorg. Would be interested to see if this happens on other desktops that don't use GTK, like KDE or something. I don't have the time to do this, but it's a start for someone else who wants this fixed more than I do (hint, hint). :)

**Comment by feren** [comment]
Yeah, BTW, DEs I've seen with the same bug:
KDE Neon-level KDE
Linux Mint-level Cinnamon
Budgie (Ubuntu Budgie version)

Given KDE is QT, GTK having the bug is out of the window entirely, though, I'm surprised that Compiz seems alright and it doesn't suffer this bug from what I've seen

**Comment by mainmachine** [comment]
OK, so it happens on desktops built on different toolkits, so lower than GTK... 

You say it doesn't happen in Compiz - in what context? Ubuntu 16.04 for example, or...?

It's looking to me like a bug in the Intel Mesa implementation - you should try using a newer Mesa stack to see if it's different in newer versions: http://www.mesa3d.org/

13.1 is the latest stable version.

**Comment by feren** [comment]
Ubuntu Unity, from what I've seen, there isn't the bug in Compiz

**Comment by s7jones** [comment]
I have this bug for Nvidia Proprietary Drivers v367.57 on Linux Mint 18. It makes it almost unbearable to log onto my Linux machine remotely using Teamviewer.

**Comment by feren** [comment]
**Has Someone Reported This To The Ubuntu Team?**

**Comment by huksley** [comment]
I`ve have had similar problems, 
solved by disabling flipping in OpenGL advanced settings in Nvidia Settings.
as per suggestion in:
https://solus-project.com/forums/viewtopic.php?t=3369

**Comment by mainmachine** [comment]
> Has Someone Reported This To The Ubuntu Team?

Have you reported it to them? Why? They don't maintain or support Teamviewer... besides, it appears as if every fix presented here has been tied back to compositing and/or vsyncing. This is a configuration issue not an application specific issue.
